### ISSUE TITLE
Message queue issues

### ISSUE THREAD

**Issue by sept-en** [trezor/trezor-core#475]
Hi, everyone!

I am currently experimenting with Trezor firmware and appropriate libraries and noticed the following problem. It seems that sometimes either device occasionally sends duplicates of messages or some messages remain in _trezord_'s queue even they should have been removed.
When it happens, there is still a chance that you will get the queue back to a normal state (i.e. you will be able to receive correct responses on your messages instead of wrong ones).

To reproduce it, I'm using _master_ branch of **trezor-core**, **python-trezor**. **Trezord-go** daemon was built from source (also _master_ branch). The tests were made mainly on Trezor T emulator but some also on mcu-emu. I experienced this error on two different machines (Ubuntu 16.04 and Arch). 

I also noticed that these things are related to Trezor's UI rendering, i.e. when something is rendered on display using built-in layout functions (e.g. _show_qr_, _confirm_response_, etc.). When such a thing happens, after any confirmation/rejection on the device's display, you either receive no response at all or wrong response message type (e.g. _Features_ instead of _Address_).

I tested on different commands built-in in trezorctl client, but I also experienced such a problem only while testing with _get_address_ and _ping -b_ commands.

Some example of the command that may rise message queue invalid behavior:
```./trezorctl get-address --address="m/44'/144'/0'/0/0" ```
And we can get this response:
```raise exceptions.TrezorException("Unexpected initial response")```
In this case, the device wasn't initialized correctly - instead of _Features_ message, the device received something that was expected to be a response for the previous command - _Address_, _Success_ etc.

We can also receive this response:
```RuntimeError: Got <class 'trezorlib.messages.Features.Features'>, expected <class 'trezorlib.messages.Address.Address'>```
So that means, that initialization message - _Features_ was sent twice (or not removed from _trezord_'s queue).

The same kind of errors could happen with other commands:
```./trezorlib ping hello -b```

I am not sure whether this is an emulator issue or daemon's one. 
And if it presents on the hardware - I haven't tested on a real device yet.

Would you give me some insights, tips or explanations on this issue, please?

P.S. I believe things described in #295 are a consequence of this issue with the message queue.

Best regards,
Kirill.

**Comment by matejcik** [comment]
are you seeing any errors on either trezorctl or the emulator, apart from the wrong messages?

trezor-core has a message queue which can come out of sync with what trezorctl/trezord is expecting. This usually happens when you hit a bug in trezorctl, which sends a message but does not retrieve the response. In that case, on your second attempt you will get a response for that previous message. Could this be the same issue?

in any case, it would be helpful if you could include an example of action that triggers this problem, even if not entirely reliably.

**Comment by sept-en** [comment]
> are you seeing any errors on either trezorctl or the emulator, apart from the wrong messages?

No

> trezor-core has a message queue which can come out of sync with what trezorctl/trezord is expecting. This usually happens when you hit a bug in trezorctl, which sends a message but does not retrieve the response. In that case, on your second attempt you will get a response for that previous message. Could this be the same issue?

Yes, it could be the same issue. Sometimes it also appears even on trying to initialize the wallet. We've tested on OSX today and have experienced it on the very first attempt.

> in any case, it would be helpful if you could include an example of action that triggers this problem, even if not entirely reliably.

I've specified some examples above (_get_address_ and _ping_ commands). It's not 100% guarantee to produce that error, though these functions are enough to test as sometimes it goes out of sync only with these.

I will have a chance to test it on the real device (Model T) next week and give some more output.

Anyway, what do you think about possibility of adding some improvements to the current message queue's mechanisms? I.e. we can add some session id for each message so we won't receive any messages from the previous commands (sessions).

**Comment by prusnak** [comment]
Redesigning the low level protocol is in the works. We plan to use either session id or message counter or both.

**Comment by matejcik** [comment]
Well, i'm unable to reproduce the problem. I'm running trezor emulator both with `./emu.sh` and with `PYOPT=0 ./emu.sh` (debug mode enabled), separately (`trezorctl -p udp:127.0.0.1:21324`) or through bridge (launched with `trezord-go -ed 21324:21325`), with or without wallet.trezor.io being open at the same time in order to hit the emulator with queries from the side.

I know that there exist _some_ sync issues, but you should not encounter them in normal operation.

So I'm wondering if there is more to it.

Could you please include _specific_ instructions that you follow to reproduce, incl. commands that you use to build the repositories, launch the emulator and bridge, any more details about your environment? Are you using unmodified sources or does this happen after you make changes? etc.

**Comment by sept-en** [comment]
@matejcik I'm using _trezor-core_ master on **38b3e44** (also tested on the latest one a few days ago). 
_python-trezor_ is on **3db0733** (latest),  _trezord-go_ is on **263c1f3** (latest).

Repositories were built following instructions in docs/readmes.
I'm launching the emu with debug option - `PYOPT=0 ./emu.sh`; bridge - `trezord-go -e 21324`.
Indeed I'm not using any specific commands/instructions to build and run.

And I can't really find the circumstances when this out of sync appears. As for now it seems truly random for me, though I continue investigating it when I have time for this. Sometimes I can work ok with it for a couple of hours and then get this behaviour for quite a long time (I believe that one out-of-sync message could break this queue). Sometimes it revives back to the normal state, sometimes I need to restart everything, including daemon.

> I know that there exist _some_ sync issues, but you should not encounter them in normal operation.

Have you encountered this issue on the real device? Or is this emulator-related only?

**Comment by matejcik** [comment]
I'm pretty sure I figured out the problem. It seems to be the bridge's fault in the emulator case, but it can happen with the real device when bridge is _not_ in use.

The message queue that is out of sync is actually on the _receiving end_ - on the USB port, or on the receiving UDP socket.
(That is why you need bridge to trigger it with an emulator - otherwise every consumer opens their own socket, which neatly sidesteps the issue)
What's happening is that at some point a consumer sends a message but then leaves the reply in the incoming buffer. From that point, all subsequent consumers will be off by one.
(with a real device OTOH, bridge's session handling will prevent this problem)

Notably, _this is not a bug of trezor-core_. The device thinks that everything is going according to protocol.

In case of emulator, restarting the bridge is all that is needed. In case of USB-connected device, the device must be reconnected, or an USB reset issued. I'll add this to python-trezor.

The following Python script triggers the issue:
```python
from trezorlib.client import get_default_client
from trezorlib.messages import Ping

c = get_default_client("bridge")  # change to "webusb" to use real device
c.transport.begin_session()
c.transport.write(Ping(button_protection=True))
c.transport.end_session()
```
After I invoke this, message queue desyncs and every subsequent `trezorctl` invocation errors out.

A similar Python script will recover from the problem, except replace `c.transport.write(...)` with `c.transport.read()`

A proper fix is session handling as per #91, but that is in the future. Fix for real devices in trezorctl is https://github.com/trezor/python-trezor/issues/361. This is unlikely to be fixed in bridge for emulator devices, because the UDP part is confusing enough as it is, and works on a best-effort basis; if we ever switch to TCP-based connection, this will magically go away.

Closing this issue, please let us know if you run into more issues that are not explained by the above.
